src/CustomElements/v1/CustomElements.js

```diff
@@ -0,0 +1,597 @@
+     * customization, upgrades and custom element reactions to be called.
+     * `enableFlush` must be true for this to work. Only use during tests!
+     */
+    flush: function() {
+      if (this.enableFlush) {
+        console.warn("flush!!!");
+        this._observers.forEach(function(observer) {
+          this._handleMutations(observer.takeRecords());
+        }, this);
+      }
+    },
+
+    _setNewInstance: function(instance) {
+      this._newInstance = instance;
+    },
+
+    /**
+     * Observes a DOM root for mutations that trigger upgrades and reactions.
+     * @private
+     */
+    _observeRoot: function(root) {
+      root.__observer = new MutationObserver(this._handleMutations.bind(this));
+      root.__observer.observe(root, {childList: true, subtree: true});
+      if (this.enableFlush) {
+        // this is memory leak, only use in tests
+        this._observers.add(root.__observer);
+      }
+    },
+
+    /**
+     * @private
+     */
+    _unobserveRoot: function(root) {
+      if (root.__observer) {
+        root.__observer.disconnect();
+        root.__observer = null;
+        if (this.enableFlush) {
+          this._observers.delete(root.__observer);
+        }
+      }
+    },
+
+    /**
+     * @private
+     */
+    _handleMutations: function(mutations) {
+      for (var i = 0; i < mutations.length; i++) {
+        var mutation = mutations[i];
+        if (mutation.type === 'childList') {
+          // Note: we can't get an ordering between additions and removals, and
+          // so might diverge from spec reaction ordering
+          this._addNodes(mutation.addedNodes);
+          this._removeNodes(mutation.removedNodes);
+        }
+      }
+    },
+
+    /**
+     * @param {NodeList} nodeList
+     * @private
+     */
+    _addNodes: function(nodeList) {
+      for (var i = 0; i < nodeList.length; i++) {
+        var root = nodeList[i];
+
+        if (!isElement(root)) {
+          continue;
+        }
+
+        // Since we're adding this node to an observed tree, we can unobserve
+        this._unobserveRoot(root);
+
+        var walker = createTreeWalker(root);
+        do {
+          var node = /** @type {HTMLElement} */ (walker.currentNode);
+          var definition = this._definitions.get(node.localName);
+          if (definition) {
+            if (!node.__upgraded) {
+              this._upgradeElement(node, definition, true);
+            }
+            if (node.__upgraded && !node.__attached) {
+              node.__attached = true;
+              if (definition && definition.connectedCallback) {
+                definition.connectedCallback.call(node);
+              }
+            }
+          }
+          if (node.shadowRoot) {
+            // TODO(justinfagnani): do we need to check that the shadowRoot
+            // is observed?
+            this._addNodes(node.shadowRoot.childNodes);
+          }
+          if (node.tagName === 'LINK') {
```

Comment:
This HTML import walk doesn't consider link de-duplication; as such, it performs an _excessive_ amount of walking through duplicated links in the dependency tree (a basic smoke test that would otherwise load 31 imports had them walked 2000+ times).  Could be mitigated by hanging a "has been walked" flag, or if needed a "last walked" monotonically increasing id off of `link.import`, and skipping links that have already been walked for this mutation, for example.